Fix mixed content error for Storage UI

globemediaofficial opened the pull request:
## What kind of change does this PR introduce?

Bug Fix

## What is the current behavior?

#11136 
#11153 
Origin of issue: https://github.com/supabase/supabase/commit/0868ee96f27333cd541dd5d512368a584a510210

The origin of the issue linked above made it so that the URL scheme was always `http`. When trying to load the storage UI from `https`, a mixed content error is shown and the buckets fail to fetch.

## What is the new behavior?

The correct scheme is now used instead of defaulting to `http`. 

## Additional context

I found that the `SUPABASE_PUBLIC_URL` in `.env` was being used to find the scheme but `constants.ts` did not have access to the variable. I changed the `.env` variable to `NEXT_PUBLIC_SUPABASE_PUBLIC_URL` so that `constants.ts` could access it and get the correct scheme. After testing locally, I can confirm that this fixes the mixed content error.

globemediaofficial committed:
Update constants.ts

globemediaofficial committed:
Update .env

globemediaofficial committed:
Update .env

alaister commented on studio/.env:
```diff
@@ -4,13 +4,13 @@ POSTGRES_PASSWORD=your-super-secret-and-long-postgres-password
 
-SUPABASE_URL=http://localhost:8000
-SUPABASE_PUBLIC_URL=http://localhost:8000
+NEXT_PUBLIC_SUPABASE_URL=http://localhost:8000
+NEXT_PUBLIC_SUPABASE_PUBLIC_URL=http://localhost:8000
```
Can we keep both to stay on the safe side? I think `SUPABASE_URL` might still be used in some API routes. For example:

```bash
SUPABASE_URL=http://localhost:8000
SUPABASE_PUBLIC_URL=http://localhost:8000
NEXT_PUBLIC_SUPABASE_URL=$SUPABASE_URL
NEXT_PUBLIC_SUPABASE_PUBLIC_URL=$SUPABASE_PUBLIC_URL
```

globemediaofficial replied:
That's probably a good idea.

globemediaofficial replied:
I made the change

alaister replied:
Thanks! What are your thoughts on using the interpolated variables (`$SUPABASE_URL`) instead of duplicating the URLs to make updating them easier?

globemediaofficial replied:
Sorry, this is new to me so I missed that in your example. I updated it to use interpolation for the `NEXT_PUBLIC` URLs.

alaister reviewed (changes requested):
Thank you @globemediaofficial! Just one small change to make

globemediaofficial committed:
Update .env to keep existing variables

globemediaofficial committed:
Update .env

alaister reviewed (approved):
Thanks again @globemediaofficial, LGTM!

globemediaofficial commented:
@alaister 
When can I expect the changes to appear on an updated docker image for the studio? I saw that @kiwicopple said the studio image is updated weekly so is there an official schedule?

serebano commented:
For docker still not fixed. Tried with latest studio image but no luck. Cannot make studio to work for https hosts

globemediaofficial commented:
> For docker still not fixed. Tried with latest studio image but no luck. Cannot make studio to work for https hosts

As far as I know, the latest image hasn’t been updated for a while. I tried to ask when it will be updated but didn’t get and answer.

pucho commented:
@globemediaofficial The docker image is being updated weekly, but we're not using the latest tag for the docker-compose file right now. We're still thinking of a good strategy for this update, in the meantime have you tried using the latest image by changing it on the docker-compose file using the latest tag [here](https://hub.docker.com/r/supabase/studio/tags?page=1)

```
studio:
    container_name: supabase-studio
    image: supabase/studio:20221214-4eecc99
```

serebano commented:
> @globemediaofficial The docker image is being updated weekly, but we're not using the latest tag for the docker-compose file right now. We're still thinking of a good strategy for this update, in the meantime have you tried using the latest image by changing it on the docker-compose file using the latest tag [here](https://hub.docker.com/r/supabase/studio/tags?page=1)
> 
> ```
> studio:
>     container_name: supabase-studio
>     image: supabase/studio:20221214-4eecc99
> ```

I already tried with supabase/studio:20230103-902aeb2

image but still no luck

pucho commented:
@globemediaofficial @serebano Would you mind providing more info on how you are setting up the https side of things for studio? I'm looking into this but can't reproduce the issue at the moment. Thanks!

globemediaofficial commented:
@pucho 

I just tried using the latest image from 3 days ago and now I have a whole separate issue. Opening the studio only shows the animated Supabase bolt and does not even show an error. Upon looking in the network tab, I see that all requests are being made to `http://localhost:8000` instead of my subdomain. I know that all of my variables in `.env` are set up correctly and they have not been changed. My setup worked as it is with the previous image (besides the storage UI because of the mixed content error). There must be a new bug which makes the studio use `http://localhost:8000`.

> Would you mind providing more info on how you are setting up the https side of things for studio? I'm looking into this but can't reproduce the issue at the moment. Thanks!
I'm using a Cloudflare tunnel with subdomains that point to ports `8443` (Kong) and `3000` (Studio). Everything worked besides the storage UI until the new issue I'm having.

pucho commented:
We introduced a small bug on the loading logic a couple of weeks ago. I just created a new docker image to fix this. 
If this works, let's move on to solving the https issues you're having.

globemediaofficial commented:
> We introduced a small bug on the loading logic a couple of weeks ago. I just created a new docker image to fix this. If this works, let's move on to solving the https issues you're having.

Everything is working for me now. This PR that I created fixed the `Mixed Content` error, so now that the other bug you found is fixed, it's working for me and the buckets load as expected.

globemediaofficial commented:
> We introduced a small bug on the loading logic a couple of weeks ago. I just created a new docker image to fix this. If this works, let's move on to solving the https issues you're having.

@pucho 
When I commented that your new docker image fixed the new issue, I had only tested it on the host computer and did not see that the Studio was still making requests to `localhost`. Today I used my subdomain to access the `Studio` on a separate computer and realized that the `localhost` requests were being made and the content cannot load. I've created a new issue #11656 for this since it is separate from the `Mixed Content` error that I fixed with this PR.

ruggi99 commented:
@alaister 
There's something wrong with this PR, IMO.

@globemediaofficial changed one line in `constants.ts` file but that change makes no sense to me.
That file is used server side (api folder) and I don't get why we are now using an environment variable with `NEXT_PUBLIC` prefix server side.

Also from what I saw `constants.ts` file is used to pass props to client side, but with this configuration it doesn't work in docker when using the image built from Supabase.
As known, `NEXT_PUBLIC` prefixed variables are evaluated at build time and so we cannot self host studio on one host and APIs on another host.

What do you think?

globemediaofficial commented:
> @alaister There's something wrong with this PR, IMO.
> 
> @globemediaofficial changed one line in `constants.ts` file but that change makes no sense to me. That file is used server side (api folder) and I don't get why we are now using an environment variable with `NEXT_PUBLIC` prefix server side.
> 
> Also from what I saw `constants.ts` file is used to pass props to client side, but with this configuration it doesn't work in docker when using the image built from Supabase. As known, `NEXT_PUBLIC` prefixed variables are evaluated at build time and so we cannot self host studio on one host and APIs on another host.
> 
> What do you think?

It was added because a different PR tried to access  `SUPABASE_PUBLIC_URL` which it could not. Adding the new variable did fix the mixed content issue because `NEXT_PUBLIC_SUPABASE_PUBLIC_URL`can be accessed. This variable was only added, the `SUPABASE_PUBLIC_URL` one is still there as @alaister requested.

ruggi99 commented:
That source is true but it doesn't mention the difference between runtime and buildtime evaluation of environment variables.
Client side variables are buildtime with a webpack plugin, server side are runtime evaluated.
Again, I'm judging the change in `constants.ts` file, not `.env` file. I think the error is there

ruggi99 commented:
https://github.com/supabase/supabase/commit/42cb1e680e1043a804f60577f1e6e2cce30076a7#comments

Also there are these comments that are right for me

globemediaofficial commented:
> That source is true but it doesn't mention the difference between runtime and buildtime evaluation of environment variables.
> Client side variables are buildtime with a webpack plugin, server side are runtime evaluated.
> Again, I'm judging the change in `constants.ts` file, not `.env` file. I think the error is there

Something may be wrong with `constants.ts`. Someone before me made a PR that made that file assume `http://localhost:8000` and I thought changing the variable would fix that. The only change I made there was switching which variable it used. I guess I’m not understanding the runtime/evaluation part.

ruggi99 commented:
I agree with this commit https://github.com/supabase/supabase/commit/0868ee96f27333cd541dd5d512368a584a510210. Looking at the code Mixed Content error should go away if you're using https in `SUPABASE_PUBLIC_URL`.
Maybe there is more going on.
Haven't tried since now, maybe I could look into this tomorrow

globemediaofficial commented:
The issue is that it didn’t work. For example, if my `SUPABASE_PUBLIC_URL` was https://supabase.example.com, the requests were made to http://supabase.example.com. That caused the mixed content error which is the whole reason I made this PR.

globemediaofficial commented:
> I agree with this commit [0868ee9](https://github.com/supabase/supabase/commit/0868ee96f27333cd541dd5d512368a584a510210). Looking at the code Mixed Content error should go away if you're using https in `SUPABASE_PUBLIC_URL`.

@ruggi99 @pucho @alaister 
### Here is what I just tried
## First
1. Set `SUPABASE_URL` and `SUPABASE_PUBLIC_URL` to `https://supabase.example.com`
2. Revert my change in `constants.ts` 
3. Run locally
    - Result was requests made to `http://supabase.example.com`, along with mixed content error

## Second
1. Set `SUPABASE_URL` and `SUPABASE_PUBLIC_URL` to `https://supabase.example.com`
2. Set `NEXT_PUBLIC_SUPABASE_PUBLIC_URL=$SUPABASE_PUBLIC_URL`
3. Use `NEXT_PUBLIC_SUPABASE_PUBLIC_URL` in  `constants.ts` 
4. Run locally
    - Result was everything working. Requests were made to `https://supabase.example.com`, no mixed content error, worked 

### What might be the issue?
A bit up in the PR, @pucho said that a logic error was introduced to the image. I wonder if there may still be an issue with the image. It seems like `SUPABASE_PUBLIC_URL` is not matching what is in `.env` on the image but it is working locally. If `SUPABASE_PUBLIC_URL` is not loading properly, `NEXT_PUBLIC_SUPABASE_PUBLIC_URL=$SUPABASE_PUBLIC_URL` will not work properly either because it will use the wrong URL.
